src/aeso_ast_infer_types.erl

```diff
@@ -1466,15 +1466,34 @@ check_reserved_entrypoints(Funs) ->
 -spec check_fundecl(env(), aeso_syntax:decl()) -> {{name(), typesig()}, aeso_syntax:decl()}.
 check_fundecl(Env, {fun_decl, Ann, Id = {id, _, Name}, Type = {fun_t, _, _, _, _}}) ->
     Type1 = {fun_t, _, Named, Args, Ret} = check_type(Env, Type),
-    {{Name, {type_sig, Ann, none, Named, Args, Ret}}, {fun_decl, Ann, Id, Type1}};
+    TypeSig = {type_sig, Ann, none, Named, Args, Ret},
+    delete_if_implementation(Env, Name, TypeSig),
+    {{Name, TypeSig}, {fun_decl, Ann, Id, Type1}};
 check_fundecl(Env, {fun_decl, Ann, Id = {id, _, Name}, Type}) ->
     type_error({fundecl_must_have_funtype, Ann, Id, Type}),
     {{Name, {type_sig, Ann, none, [], [], Type}}, check_type(Env, Type)}.
 
+%% Delete the function FunName with the signature FunSig from of functions
+%% to be implemented if it is included there, or return ok otherwise.
+-spec delete_if_implementation(env(), FunName, FunSig) -> true when
```

Comment:
I don't like this name. It is not clear from where you delete (the universe?). At first I thought you delete it from the contract. Why not something like "register implementation"?

Reply:
Changed here 06879fdf7384fff7c038196a2f5799aefdfcd98a